src/btree/col_srch.c

```diff
@@ -9,6 +9,53 @@
 #include "wt_internal.h"
 
 /*
+ * __check_leaf_key_range --
+ *	Check the search key is in the leaf page's key range.
+ */
+static inline int
+__check_leaf_key_range(WT_SESSION_IMPL *session,
+    uint64_t recno, WT_REF *leaf, WT_CURSOR_BTREE *cbt)
+{
+	WT_PAGE_INDEX *pindex;
+	uint32_t indx;
+
+	/*
+	 * Check if the search key is less than the parent's starting key for
+	 * this page.
+	 */
+	if (recno < leaf->key.recno) {
+		cbt->compare = 1;		/* page keys > search key */
+		return (0);
+	}
+
+	/*
+	 * Check if the search key is greater than or equal to the starting key
+	 * for the parent's next page.
+	 *
+	 * !!!
+	 * Check that "indx + 1" is a valid page-index entry first, because it
+	 * also checks that "indx" is a valid page-index entry, and we have to
+	 * do that latter check before looking at the indx slot of the array
+	 * for a match to leaf (in other words, our page hint might be wrong).
+	 */
+	WT_INTL_INDEX_GET(session, leaf->home, pindex);
+	indx = leaf->pindex_hint;
+	if (indx + 1 < pindex->entries && pindex->index[indx] == leaf)
```

Comment:
> Should the test in the first clause be <=? I thought the pindex_hint field was zero-based.

I think it's correct as written. The idea is we're checking two fields, `indx` and `indx + 1`, and both of them have to be less than `pindex->entries`. Instead of doing two tests, I'm doing one test, for the larger value.

It is zero-based, so if we're going to look at slots 3 and 4, then `pindex->entries` must be at least 5.

Please let me know if this doesn't make sense, and we'll talk it over.